src/inline-assembly.md

```diff
@@ -39,513 +54,964 @@ assert_eq!(x, 4 * 6);
+> | ARM | `qreg` | None | `q0` | `q` |
+> | ARM | `qreg` | `e` / `f` | `d0` / `d1` | `e` / `f` |
+> | RISC-V | `reg` | None | `x1` | None |
+> | RISC-V | `freg` | None | `f0` | None |
+> | LoongArch | `reg` | None | `$r1` | None |
+> | LoongArch | `freg` | None | `$f0` | None |
 
-> **Notes**:
+```rust
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32 = 5;
+let y: i32;
+core::arch::asm!("mov {:e}, {:e}", out(reg) y, in(reg) x);
+# }}
+```
+
+> [!NOTE]
+> The supported modifiers are a subset of LLVM's (and GCC's) [asm template argument modifiers][llvm-argmod], but do not use the same letter codes.
+
+> [!NOTE]
 > - on ARM `e` / `f`: this prints the low or high doubleword register name of a NEON quad (128-bit) register.
 > - on x86: our behavior for `reg` with no modifiers differs from what GCC does.
 >   GCC will infer the modifier based on the operand value type, while we default to the full register size.
 > - on x86 `xmm_reg`: the `x`, `t` and `g` LLVM modifiers are not yet implemented in LLVM (they are supported by GCC only), but this should be a simple change.
 
-As stated in the previous section, passing an input value smaller than the register width will result in the upper bits of the register containing undefined values.
-This is not a problem if the inline asm only accesses the lower bits of the register, which can be done by using a template modifier to use a subregister name in the asm code (e.g. `ax` instead of `rax`).
-Since this an easy pitfall, the compiler will suggest a template modifier to use where appropriate given the input type.
-If all references to an operand already have modifiers then the warning is suppressed for that operand.
+r[asm.template.diagnostic]
+A lint diagnostic should be emitted if a modifier is omitted, or a modifier is used such that the modified expanded register is of an inappropriate width for the type used to initialize the operand
 
 [llvm-argmod]: http://llvm.org/docs/LangRef.html#asm-template-argument-modifiers
 
-## ABI clobbers
+## Behaviour of an asm block
 
-The `clobber_abi` keyword can be used to apply a default set of clobbers to an `asm!` block.
-This will automatically insert the necessary clobber constraints as needed for calling a function with a particular calling convention: if the calling convention does not fully preserve the value of a register across a call then `lateout("...") _` is implicitly added to the operands list (where the `...` is replaced by the register's name).
+r[asm.evaluation]
 
-`clobber_abi` may be specified any number of times. It will insert a clobber for all unique registers in the union of all specified calling conventions.
+r[asm.evaluation.general]
+Each evaluation of an *asm block* shall perform an operation that correpsonds to the result of a valid sequence of operations on the Minirust Abstract Machine. The behaviour is undefined if the operations performed by the asm block do not validly correspond to a valid sequence of Minirust operations.
```

Comment:
This seems like a new claim that we will be defining things in terms of Minirust. Is that something we have decided to do?

Reply:
So there are two separate "new claims" here - one is that the operational semantics will be defined in terms of Minirust in general (that had a whole meeting on it, where we took an hour and a half to say "Yes"... actually, no, we took 5 minutes to say "Yes" and then spent the other hour 25 minutes debating "to what extent"), and the other is that inline assembly must affect the Rust AM in a way that Rust Code could possibly have done.

I believe the latter has been the opinion of the opsem team since before it was the opsem team, and is more generally, makes the claim that compilers can optimize Rust in the face of unanalyzed code (which has been assumed as true since we started plugging Rust into llvm).

Reply:
In practice, this means that this code has UB, because it doesn't correspond back to a valid sequence of Minirust operations (that do not themselves have undefined behaviour):
```rust
let x: i32 = 0;
asm!("mov dword ptr [{}], 5", in(reg) &x);
println!("{x}"); // May be turned into `println!("0");`
```

The code writes to a `Frozen` tag, which is UB. But if inline assembly isn't constrained to simply doing things (to the AM) Rust code itself can do, then it could circumvent that rule.

Reply:
> that had a whole meeting on it, where we took an hour and a half to say "Yes"... actually, no, we took 5 minutes to say "Yes" and then spent the other hour 25 minutes debating "to what extent")

I don't recall having a team decision on this point, and I don't see any discussions about that decision in the [meeting notes](https://hackmd.io/team/rust-spec-team?nav=overview). 

I'm not opposed to Minirust as an eventual basis to use for defining things, but I'm not yet convinced that we are in a position where we are ready to do that *today*.

I'd also say that adding things like this are pretty heavy decisions that are probably best not to bundle in one large update. This makes it harder to review and make progress when there are many new changes included in one PR. Is it possible to work in smaller batches, and to keep new additions separate from things that are more editorially oriented?

Reply:
I think we need to be able to say what an inline assembly block does, and I'm not sure there's any other model we can use. We know that it can do more than Rust code can.  For example, an asm block can yield a nondeterministic value, directly accessing the `pick` Minirust operation.

Reply:
@ehuss is right here.  Maybe we do...

> need to be able to say what an inline assembly block does

...in more detail than what the Reference chapter says today.  But that doesn't need to all happen in this PR.  And trying to do that is making it difficult to review this and make progress here.

Reply:
FWIW this paragraph is not specific to inline assembly, it generally applies to any kind of FFI call.

Reply:
Clarifying my statement at the T-spec meeting:
* I think we need *a* claim here, for the rest of the section and the following section ([asm.options]) to make sense.
* I don't think it necessarily needs to make a claim about Minirust.
* We know that the claim should allow *something* more than what Rust itself allows, but less than what the hardware allows.